packages/navbrand/src/react/index.tsx

```diff
@@ -34,52 +38,79 @@ const styles = {
     )
 }
 
-const NavBrand = React.forwardRef((props, forwardedRef) => {
-  const ref = React.useRef()
-  React.useImperativeHandle(forwardedRef, () => ref.current)
+interface BaseNavBrandProps {
+  logo?: ReactNode
+  wordmark?: ReactNode
+}
+interface AnchorProps extends BaseNavBrandProps, PropsOf<'a'> {
+  href: string
+  onClick?: MouseEventHandler<HTMLAnchorElement>
+}
+interface ButtonProps extends BaseNavBrandProps, PropsOf<'button'> {
+  href?: undefined
+  onClick: MouseEventHandler<HTMLButtonElement>
+}
+interface DivProps extends BaseNavBrandProps, PropsOf<'div'> {
+  href?: undefined
+  onClick?: MouseEventHandler<HTMLDivElement>
+}
 
-  const tagName = props.href ? 'a' : props.onClick ? 'button' : 'div'
+type NavBrandElement = HTMLAnchorElement | HTMLButtonElement | HTMLDivElement
+type NavBrandProps = AnchorProps | ButtonProps | DivProps
+type NavBrandComponent = {
+  (props: AnchorProps, ref?: RefFor<'a'>): JSX.Element
+  (props: ButtonProps, ref?: RefFor<'button'>): JSX.Element
+  (props: DivProps, ref?: RefFor<'div'>): JSX.Element
+}
 
-  return (
-    <Halo inline gapSize={Halo.gapSizes.small}>
-      {React.createElement(
-        tagName,
-        {
-          ...styles.navBrand(props),
-          ...filterReactProps(props, { tagName }),
-          ref
-        },
-        <>
-          <Logo {...props} />
-          <Wordmark {...props} />
-        </>
-      )}
-    </Halo>
-  )
-})
+const NavBrand = React.forwardRef<NavBrandElement, NavBrandProps>(
+  (props, forwardedRef) => {
+    const { logo, wordmark, ...rest } = props
 
-NavBrand.displayName = 'NavBrand'
-NavBrand.propTypes = {
-  href: PropTypes.string,
-  logo: PropTypes.node,
-  onClick: PropTypes.func,
-  wordmark: PropTypes.node
-}
+    const ref = React.useRef<NavBrandElement>(null)
+    React.useImperativeHandle(
+      forwardedRef,
+      () => (ref.current as unknown) as NavBrandElement
+    )
+    const isAnchor = 'href' in props
+    const isButton = !isAnchor && 'onClick' in props
+
+    const Wrapper: React.FC = wrapperProps =>
+      isAnchor ? (
+        <a
+          ref={ref as RefFor<'a'>}
+          {...(rest as PropsOf<'a'>)}
+          {...wrapperProps}
+        />
+      ) : isButton ? (
+        <button
+          ref={ref as RefFor<'button'>}
+          {...wrapperProps}
+          {...(rest as PropsOf<'button'>)}
+        />
+      ) : (
+        <div
+          ref={ref as RefFor<'div'>}
+          {...wrapperProps}
+          {...(rest as PropsOf<'div'>)}
+        />
+      )
+
+    return (
+      <Halo inline gapSize={Halo.gapSizes.small}>
+        <Wrapper {...styles.navBrand(props)}>
+          <Logo>{logo}</Logo>
+          <Wordmark>{wordmark}</Wordmark>
```

Comment:
What was your reasoning for changing these internal apis?

Reply:
two fold.
  - limit what is passed to them as props. just children instead of all nav brand props
  - use children because it's standard and doesn't need additional typings